File: pixel_sort.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

int main(int argc, char** argv) {

    if (argc != 2) {
        cerr << "Usage: ./pixelsort <image_file>" << endl;
        return 1;
    }

    // Load input image
    Mat image = imread(argv[1]);
    if (image.empty()) {
        cerr << "Error: Could not load image " << argv[1] << endl;
        return 1;
    }

    // Define parameters for pixel sorting
    int channel = 0; // Use the first color channel (BGR order)
    bool sort_rows = true; // Sort rows instead of columns
    int start_row = 100; // Only sort rows starting from row 100
    int end_row = 200; // Only sort rows up to row 200

    // Extract rows to be sorted
    Mat rows_to_sort = image(Range(start_row, end_row + 1), Range::all());

    // Convert rows to grayscale
    Mat gray_rows;
    cvtColor(rows_to_sort, gray_rows, COLOR_BGR2GRAY);

    // Apply pixel sorting to grayscale rows
    Mat sorted_rows;
    sortIdx(gray_rows.reshape(1, 0), sorted_rows, SORT_EVERY_ROW + SORT_ASCENDING);

    // Create output image with sorted rows
    Mat sorted_image = image.clone();
    for (int r = 0; r < sorted_rows.rows; r++) {
        int sorted_row_idx = sorted_rows.at<int>(r, 0);
        Mat sorted_row = rows_to_sort.row(sorted_row_idx).clone();
        sorted_row.copyTo(sorted_image.row(start_row + r));
    }

// Save sorted image
imwrite(argv[1], sorted_image);

return 0;

}

File: pixel_sort.py
from PIL import Image

# Load input image
input_file = 'input.jpg'
input_image = Image.open(input_file)

# Define parameters for pixel sorting
sort_rows = True # Sort rows instead of columns
start_row = 100 # Only sort rows starting from row 100
end_row = 200 # Only sort rows up to row 200

# Extract rows or columns to be sorted
if sort_rows:
    rows_to_sort = input_image.crop((0, start_row, input_image.width, end_row + 1))
    pixels_to_sort = list(rows_to_sort.getdata())
else:
    columns_to_sort = input_image.crop((start_row, 0, end_row + 1, input_image.height))
    pixels_to_sort = list(columns_to_sort.getdata())

# Sort pixels
sorted_pixels = sorted(pixels_to_sort)

# Create output image with sorted pixels
sorted_image = input_image.copy()
if sort_rows:
    sorted_rows = Image.new('RGB', (input_image.width, end_row - start_row + 1))
    sorted_rows.putdata(sorted_pixels)
    sorted_image.paste(sorted_rows, (0, start_row))
else:
    sorted_columns = Image.new('RGB', (end_row - start_row + 1, input_image.height))
    sorted_columns.putdata(sorted_pixels)
    sorted_image.paste(sorted_columns, (start_row, 0))

#Save sorted image

output_file = 'output.jpg'
sorted_image.save(output_file)